src/webui/webapplication.cpp

```diff
@@ -893,98 +823,133 @@ void WebApplication::action_command_setCategory()
         BitTorrent::TorrentHandle *const torrent = BitTorrent::Session::instance()->findTorrent(hash);
         if (torrent) {
             if (!torrent->setCategory(category)) {
-                status(400, "Incorrect category name");
+                status(404, "Not Found");
+                print(tr("Incorrect category name"), Http::CONTENT_TYPE_TXT);
                 return;
             }
         }
     }
 }
 
-void WebApplication::action_command_addCategory()
+void WebApplication::addCategory()
 {
-    CHECK_URI(0);
     CHECK_PARAMETERS("category");
 
     QString category = request().posts["category"].trimmed();
 
     if (!BitTorrent::Session::isValidCategoryName(category) && !category.isEmpty()) {
-        status(400, tr("Incorrect category name"));
+        status(404, "Not Found");
+        print(tr("Incorrect category name"), Http::CONTENT_TYPE_TXT);
         return;
     }
 
     BitTorrent::Session::instance()->addCategory(category);
 }
 
-void WebApplication::action_command_removeCategories()
+void WebApplication::removeCategories()
 {
-    CHECK_URI(0);
     CHECK_PARAMETERS("categories");
 
     QStringList categories = request().posts["categories"].split('\n');
     foreach (const QString &category, categories)
         BitTorrent::Session::instance()->removeCategory(category);
 }
 
-void WebApplication::action_command_getSavePath()
+void WebApplication::getDefaultSavePath()
 {
-    CHECK_URI(0);
     print(BitTorrent::Session::instance()->defaultSavePath());
 }
 
-bool WebApplication::isPublicScope()
+bool WebApplication::isAPIMethod(const QString &methodName) const
 {
-    return (scope_ == DEFAULT_SCOPE || scope_ == VERSION_INFO);
+    return (m_privateAPIs.contains(methodName)
+            || m_publicAPIs.contains(methodName));
+}
+
+bool WebApplication::isPublicAPIMethod(const QString &methodName) const
+{
+    return m_publicAPIs.contains(methodName);
 }
 
 void WebApplication::doProcessRequest()
 {
-    scope_ = DEFAULT_SCOPE;
-    action_ = DEFAULT_ACTION;
+    if (request().path.startsWith(PATH_PREFIX_API)) {
+        const QString methodName {request().path.mid(PATH_PREFIX_API.size())};
+        if (!isAPIMethod(methodName)) {
+            status(404, "Not Found");
+            return;
+        }
 
-    parsePath();
+        if (!sessionActive() && !isPublicAPIMethod(methodName)) {
+            status(403, "Forbidden");
+            return;
+        }
 
-    if (args_.contains(".") || args_.contains("..")) {
-        qDebug() << Q_FUNC_INFO << "Invalid path:" << request().path;
-        status(404, "Not Found");
-        return;
+        QMetaObject::invokeMethod(this, methodName.toLatin1().constData());
```

Comment:
I believe parameters marshaling can be done here as well. Motivation: API functions would contain parameters in its signatures (documenting themselves) and Remote API would be one step away from becoming protocol-agnostic.

Reply:
>I believe parameters marshaling can be done here as well. Motivation: API functions would contain parameters in its signatures (documenting themselves) and Remote API would be one step away from becoming protocol-agnostic.

I also think about it. But there are some difficulties here. Let me think about it after this PR.

Reply:
What kind of difficulties?

Reply:
I don't think unsolvable. I have already tried something.
But it is still beyond the scope of this PR, so I would prefer to do it separately.

Reply:
Besides, to make it really protocol-agnostic, we need to implement additional abstraction layer. Need to develop some requirements (e.g. the format of the input and output data, status, etc.).